{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e7bf8e13",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import json\n",
    "import os\n",
    "\n",
    "from itertools import permutations\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9d7b6fa5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1711760</td>\n",
       "      <td>Delayed institution of hypertension during foc...</td>\n",
       "      <td>The effect of induced hypertension instituted ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6086495</td>\n",
       "      <td>Localisation of the Becker muscular dystrophy ...</td>\n",
       "      <td>A linkage study in 30 Becker muscular dystroph...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7018927</td>\n",
       "      <td>Pituitary response to luteinizing hormone-rele...</td>\n",
       "      <td>The effects of a 6-hour infusion with haloperi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7811247</td>\n",
       "      <td>X-linked adrenoleukodystrophy (ALD): a novel m...</td>\n",
       "      <td>Fragments of the adrenoleukodystrophy (ALD) cD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8944024</td>\n",
       "      <td>Detection of heterozygous mutations in BRCA1 u...</td>\n",
       "      <td>The ability to scan a large gene rapidly and a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   abstract_id                                              title  \\\n",
       "0      1711760  Delayed institution of hypertension during foc...   \n",
       "1      6086495  Localisation of the Becker muscular dystrophy ...   \n",
       "2      7018927  Pituitary response to luteinizing hormone-rele...   \n",
       "3      7811247  X-linked adrenoleukodystrophy (ALD): a novel m...   \n",
       "4      8944024  Detection of heterozygous mutations in BRCA1 u...   \n",
       "\n",
       "                                            abstract  \n",
       "0  The effect of induced hypertension instituted ...  \n",
       "1  A linkage study in 30 Becker muscular dystroph...  \n",
       "2  The effects of a 6-hour infusion with haloperi...  \n",
       "3  Fragments of the adrenoleukodystrophy (ALD) cD...  \n",
       "4  The ability to scan a large gene rapidly and a...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstracts_test_fname = \"data/abstracts_test.csv\"\n",
    "abstracts_test_df = pd.read_csv(abstracts_test_fname,sep='\\t')\n",
    "abstracts_test_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0a72e48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>abstract_id</th>\n",
       "      <th>offset_start</th>\n",
       "      <th>offset_finish</th>\n",
       "      <th>type</th>\n",
       "      <th>mention</th>\n",
       "      <th>entity_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1711760</td>\n",
       "      <td>23</td>\n",
       "      <td>35</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>hypertension</td>\n",
       "      <td>D006973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1711760</td>\n",
       "      <td>49</td>\n",
       "      <td>66</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>cerebral ischemia</td>\n",
       "      <td>D002545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1711760</td>\n",
       "      <td>78</td>\n",
       "      <td>89</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>brain edema</td>\n",
       "      <td>D001929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1711760</td>\n",
       "      <td>113</td>\n",
       "      <td>125</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>hypertension</td>\n",
       "      <td>D006973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1711760</td>\n",
       "      <td>165</td>\n",
       "      <td>197</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>middle cerebral artery occlusion</td>\n",
       "      <td>D020244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id  abstract_id  offset_start  offset_finish                        type  \\\n",
       "0   0      1711760            23             35  DiseaseOrPhenotypicFeature   \n",
       "1   1      1711760            49             66  DiseaseOrPhenotypicFeature   \n",
       "2   2      1711760            78             89  DiseaseOrPhenotypicFeature   \n",
       "3   3      1711760           113            125  DiseaseOrPhenotypicFeature   \n",
       "4   4      1711760           165            197  DiseaseOrPhenotypicFeature   \n",
       "\n",
       "                            mention entity_ids  \n",
       "0                      hypertension    D006973  \n",
       "1                 cerebral ischemia    D002545  \n",
       "2                       brain edema    D001929  \n",
       "3                      hypertension    D006973  \n",
       "4  middle cerebral artery occlusion    D020244  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities_test_fname = \"data/entities_test.csv\"\n",
    "entities_test_df = pd.read_csv(entities_test_fname,sep='\\t')\n",
    "entities_test_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "02d85ecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_id2title ={ abs_id: title for abs_id, title in abstracts_test_df[['abstract_id','title']].values }\n",
    "test_id2abstr ={ abs_id: abs_text for abs_id, abs_text in abstracts_test_df[['abstract_id','abstract']].values }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8ddd44f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "3263\n"
     ]
    }
   ],
   "source": [
    "print(len(abstracts_test_df))\n",
    "print(len(entities_test_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "18ac78b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# entity-type dictionary\n",
    "entityid2type ={ entity_ids: type for entity_ids, type in entities_test_df[['entity_ids','type']].values }\n",
    "mention2type = { mention: type for mention, type in entities_test_df[['mention','type']].values }\n",
    "mention2entityid = { mention: entity_ids for mention, entity_ids in entities_test_df[['mention','entity_ids']].values }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2d0588e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "entity_type_valid = [('GeneOrGeneProduct', 'GeneOrGeneProduct'),\n",
    "                     ('GeneOrGeneProduct', 'DiseaseOrPhenotypicFeature'),\n",
    "                     ('ChemicalEntity', 'DiseaseOrPhenotypicFeature'),\n",
    "                     ('DiseaseOrPhenotypicFeature', 'GeneOrGeneProduct'),\n",
    "                     ('SequenceVariant', 'DiseaseOrPhenotypicFeature'),\n",
    "                     ('ChemicalEntity', 'GeneOrGeneProduct'),\n",
    "                     ('DiseaseOrPhenotypicFeature', 'SequenceVariant'),\n",
    "                     ('DiseaseOrPhenotypicFeature', 'ChemicalEntity'),\n",
    "                     ('ChemicalEntity', 'ChemicalEntity'),\n",
    "                     ('GeneOrGeneProduct', 'ChemicalEntity'),\n",
    "                     ('SequenceVariant', 'ChemicalEntity'),\n",
    "                     ('ChemicalEntity', 'SequenceVariant'),\n",
    "                     ('SequenceVariant', 'SequenceVariant'),\n",
    "                     ('SequenceVariant', 'GeneOrGeneProduct'),\n",
    "                     ('GeneOrGeneProduct', 'SequenceVariant')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "68befac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### generete test dataset file\n",
    "sentence2entitise = defaultdict(list)\n",
    "sentence_infos = dict()\n",
    "relation_candidates = []\n",
    "for key, df in entities_test_df.groupby('abstract_id'):\n",
    "    #if key == 1711760:\n",
    "    text = test_id2title[key] + ' ' + test_id2abstr[key]\n",
    "    sentences = text.split('. ')\n",
    "    entity_infos = dict()\n",
    "    mention2entityid = {}\n",
    "    sen2ents = {}\n",
    "    for i, row in df.iterrows():\n",
    "        entity_infos[row['offset_start'], row['offset_finish']] = row['mention']\n",
    "        mention2entityid[row['mention']] = row['entity_ids']\n",
    "        #entity = row['mention']\n",
    "\n",
    "    for idx in range(len(sentences)):\n",
    "        if idx == 0:\n",
    "            sent_left = 0\n",
    "            sent_right = len(sentences[idx]+'. ')\n",
    "        else:\n",
    "            sent_left = sent_right #+ 1\n",
    "            sent_right += len(sentences[idx]+'. ')\n",
    "        for entity_pos in entity_infos.keys():\n",
    "            if entity_pos[0] > sent_left and entity_pos[1] < sent_right:\n",
    "                sentence2entitise[sentences[idx]].append(entity_infos[entity_pos])\n",
    "\n",
    "    for s in sentences:\n",
    "        entities =[]\n",
    "        for m in mention2entityid:\n",
    "            if m in s:\n",
    "                entities.append(m)\n",
    "\n",
    "        entities = list(set(entities))\n",
    "        comb = permutations(entities, 2)\n",
    "        for c in comb:\n",
    "            if (mention2type[c[0]], mention2type[c[1]]) in entity_type_valid:\n",
    "                entity_1_mention, entity_2_mention = c[0], c[1]\n",
    "                entity_1_id, entity_2_id = mention2entityid[c[0]], mention2entityid[c[1]]\n",
    "                relation_candidates.append([key, \n",
    "                                            s, \n",
    "                                            entity_1_mention, \n",
    "                                            entity_2_mention, \n",
    "                                            entity_1_id, \n",
    "                                            entity_2_id])\n",
    "                # if in entity_type_valid\n",
    "                #  else: \n",
    "                # if we do not remove the entity-entity which not in entity_type_valid\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "be9ad9bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "size of test data:  8480\n"
     ]
    }
   ],
   "source": [
    "dataset_df =pd.DataFrame(relation_candidates, \n",
    "                         columns=['abstract_id','sentence','entity_1_mention','entity_2_mention','entity_1_id','entity_2_id'])\n",
    "print('size of test data: ',len(dataset_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eb3f3365",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract_id</th>\n",
       "      <th>sentence</th>\n",
       "      <th>entity_1_mention</th>\n",
       "      <th>entity_2_mention</th>\n",
       "      <th>entity_1_id</th>\n",
       "      <th>entity_2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6879</th>\n",
       "      <td>26516699</td>\n",
       "      <td>Interestingly, the differential expression of ...</td>\n",
       "      <td>Let-7b</td>\n",
       "      <td>MSST1</td>\n",
       "      <td>406884</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4212</th>\n",
       "      <td>22048266</td>\n",
       "      <td>Significantly lower frequency of SIM2 C-G hapl...</td>\n",
       "      <td>rs2073416</td>\n",
       "      <td>DS</td>\n",
       "      <td>rs2073416</td>\n",
       "      <td>D004314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1585</th>\n",
       "      <td>17318851</td>\n",
       "      <td>The same variants in the IRF6 gene that are as...</td>\n",
       "      <td>orofacial clefts</td>\n",
       "      <td>rs861019</td>\n",
       "      <td>C566121</td>\n",
       "      <td>rs861019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6621</th>\n",
       "      <td>26516699</td>\n",
       "      <td>In addition, from 10 to 50 weeks of age, stage...</td>\n",
       "      <td>Let-96</td>\n",
       "      <td>miR-195</td>\n",
       "      <td>-</td>\n",
       "      <td>387190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5456</th>\n",
       "      <td>24853300</td>\n",
       "      <td>Immunohistochemistry showed that CBKOTg mice h...</td>\n",
       "      <td>CB</td>\n",
       "      <td>amyloid b-peptide</td>\n",
       "      <td>12307</td>\n",
       "      <td>11820</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      abstract_id                                           sentence  \\\n",
       "6879     26516699  Interestingly, the differential expression of ...   \n",
       "4212     22048266  Significantly lower frequency of SIM2 C-G hapl...   \n",
       "1585     17318851  The same variants in the IRF6 gene that are as...   \n",
       "6621     26516699  In addition, from 10 to 50 weeks of age, stage...   \n",
       "5456     24853300  Immunohistochemistry showed that CBKOTg mice h...   \n",
       "\n",
       "      entity_1_mention   entity_2_mention entity_1_id entity_2_id  \n",
       "6879            Let-7b              MSST1      406884           -  \n",
       "4212         rs2073416                 DS   rs2073416     D004314  \n",
       "1585  orofacial clefts           rs861019     C566121    rs861019  \n",
       "6621            Let-96            miR-195           -      387190  \n",
       "5456                CB  amyloid b-peptide       12307       11820  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fcdf95d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_df.to_csv('./data/test_data2.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee3841cb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
